static/js/util/coupon.js

```diff
@@ -0,0 +1,85 @@
+// @flow
+import {
+  COUPON_CONTENT_TYPE_PROGRAM,
+  COUPON_CONTENT_TYPE_COURSE,
+  COUPON_CONTENT_TYPE_COURSERUN,
+  COUPON_AMOUNT_TYPE_PERCENT_DISCOUNT,
+  COUPON_AMOUNT_TYPE_FIXED_DISCOUNT,
+} from '../constants';
+import type {
+  Coupons,
+  Coupon,
+  CalculatedPrices,
+} from '../flow/couponTypes';
+import type {
+  CoursePrice,
+  CoursePrices,
+  Dashboard,
+} from '../flow/dashboardTypes';
+
+export const calculatePrices = (programs: Dashboard, prices: CoursePrices, coupons: Coupons): CalculatedPrices => {
+  let couponLookup: Map<number, Coupon> = new Map();
+  let priceLookup: Map<number, CoursePrice> = new Map();
+  for (const coupon of coupons) {
+    couponLookup.set(coupon.program_id, coupon);
+  }
+  for (const price of prices) {
+    priceLookup.set(price.program_id, price);
+  }
```

Comment:
it might be nice to pull these two loops out into some helper functions, something like:

```js
const populateMap = R.curry((getKey, objArray) => {
  let map = new Map();
  objArray.forEach(obj => map.set(getKey(obj), obj));
});

const makeCouponMap: (c: Coupons) => Map<number, Coupon> = populateMap(
  R.prop('program_id'))
);

const makePriceMap: (c: Coupons) => Map<number, Coupon> = populateMap(
  R.prop('program_id'))
);
```

hum just realized both of those are using the `program_id` prop, but you'd probably still want to annotate the return types separately, maybe you could just do:

```js
const makeProgramIDMap = populateMap(R.prop('program_id'));

let couponLookup: Map<number, Coupon> = makeProgramIDMap(coupons);
let priceLookup: Map<number, Coupon> = makeProgramIDMap(prices);
```

just a thought!